### ISSUE TITLE
InputControl and TextareaControl: label, placeholder, and background colors should work consistently and be always readable

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#69355]
### Description

While going through https://github.com/WordPress/gutenberg/issues/60275, https://github.com/WordPress/gutenberg/pull/68807, https://github.com/WordPress/gutenberg/pull/69312, and https://github.com/WordPress/gutenberg/pull/69334, I noticed `InputControl` and `TextareaControl` hav einconsistent and buggy behaviors for:
- The label color.
- The placeholder attribute color.
- The input and textarea background color.
- The input and textarea background color when `disabled`.

This is more noticeable when using the components inside the editor canvas with both a light and dark backgrounds.

One of these buggy behaviors is surfaced after the `is-dark-theme` CSS class for the Post Editor was fixed in https://github.com/WordPress/gutenberg/pull/68481

I'd expect these two components to use a consistent styling for all states and have the same behavior with a dark background theme.
Cc @WordPress/gutenberg-components 

To quickly test, I added the components below to [the PostTitle](https://github.com/WordPress/gutenberg/blob/fix/placeholder-text-contrast/packages/editor/src/components/post-title/index.js/#L172):

<details><summary>Added components!</summary>

```
<TextareaControl
	label="TextareaControl"
	placeholder="Placeholder"
	rows={ 3 }
	__nextHasNoMarginBottom
/>
<br />
<InputControl
	__next40pxDefaultSize
	type="text"
	value=""
	label="InputControl"
	placeholder="Placeholder"
/>
<br />
<TextareaControl
	label="TextareaControl disabled"
	placeholder="Placeholder"
	rows={ 3 }
	__nextHasNoMarginBottom
	disabled
/>
<br />
<InputControl
	__next40pxDefaultSize
	type="text"
	value=""
	label="InputControl disabled"
	placeholder="Placeholder"
	disabled
/>
<br />
```

</details>

|Light|Dark|
|-|-|
|![Image](https://github.com/user-attachments/assets/fbc033ab-df0c-484d-9ea9-ff28149f1148)|![Image](https://github.com/user-attachments/assets/f6b87d84-a799-4b2a-bff7-8d4f687feb99)|

Once the ongoing work on the placeholder text color in the issues linked above is complete, I'll try to raise a PR for these additional fixes.

### Step-by-step reproduction instructions

- Add the components as explained above.
- Edit a post. Compare with a theme with light background and another one with dark background (e.g. Twenty Twenty-Four 'Maelstrom' variation).
- Observe that:
- TextareaControl for both enabled and disabled state doesn't have a background color set. It uses the browser default background color.
- InputControl does have a background color but it comes from the container, while the input has a transparent color.
- With a dark theme:
  - The TextareaControl placeholder is barely visible. Expected: to stay dark gray.
  - The InputControl label color doesn't switch to white. The TextareaControl does.
  - The TextareaControl background color for the disabled state is the browser default color.
  - The InputControl background color for the disabled state is a very light gray.

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [ ] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

**Comment by im3dabasia** [comment]
Thanks for raising this issue. For additional context, related changes were introduced in this PR: https://github.com/WordPress/gutenberg/pull/68349. There was further discussion here: https://github.com/WordPress/gutenberg/pull/68349#discussion_r1900603813 regarding the use of the `optimizeReadabilityFor` prop.

In that discussion, it was agreed that we can use `optimizeReadabilityFor` prop to fix the text color issue.

> For example, when using `<Text optimizeReadabilityFor="#000">`, the text color must resolve to white for readability, regardless of the active theme colors.


Related PRs: https://github.com/WordPress/gutenberg/pull/68349, https://github.com/WordPress/gutenberg/pull/68472

Looping in @mirka for your thoughts.

**Comment by mirka** [comment]
Is there an actual legitimate usage in Gutenberg where this problem manifests? Because the experimental theming features in `@wordpress/components` are not officially supported, and will be superseded by an official theming system in `@wordpress/ui`.

Asking because the screenshot in the OP seems to place `@wordpress/components` in a front-end context (i.e. directly in the block editor writing flow), something that isn't supported as far as I know.

**Comment by mirka** [comment]
Closing because the use cases are out of scope.
